gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,267 @@
+
+        # Set image dimensions to those of zero’th image
+        self._h, self._w = (
+            self._images[self._patchmatchnet_idx_to_camera_idx[0]].height,
+            self._images[self._patchmatchnet_idx_to_camera_idx[0]].width,
+        )
+        # Calculate the cropped size of each image so that the height and width of the image can be divided by 8 evenly,
+        #   so that the image size can be matched after downsampling and then upsampling in PatchmatchNet
+        self._cropped_h, self._cropped_w = (self._h - self._h % 8, self._w - self._w % 8)
+
+        self._pairs, self._depth_ranges = self.configure()
+
+    def configure(self) -> Tuple[np.ndarray, np.ndarray]:
+        """Configure pairs and depth_ranges for each view from sfm_result.
+        Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/colmap_input.py (line 360-410)
+
+        If there are N0 valid images and the patchmatchnet's number of views is num_views, the function does:
+            1. Calculate the scores between each pair of the N0 images. Then for every image as the reference image,
+            find (num_views - 1) images with highest scores as the source images.
+            2. The scores between each pair of the N0 images is the sum of piecewise Gaussian scores between vectors
+            from a common track's coordinates to each camera center, which is mentioned in "View Selection" paragraphs
+            in Yao's paper https://arxiv.org/abs/1804.02505.
+            3. For every image as the reference image, calculate the depth range.
+
+        Returns:
+            pairs: array of shape (num_images, num_views-1). Each row_id indicates the index of reference view
+                in self.keys, with (num_views-1) values indicating the indices of source views in self.keys
+            depth_ranges: array of shape (num_images, 2). Each row_id indicates the index of reference view
+                in self.keys, with 2 values indicating [min_depth, max_depth]
+        """
+        num_tracks = self._sfm_result.number_tracks()
+
+        # Initialize the pairwise scores between the same views as negative infinity
+        pair_scores = np.zeros((self._num_images, self._num_images))
+        np.fill_diagonal(pair_scores, -np.inf)
+
+        # Initialize empty lists to collect all possible depths for each view
+        depths: List[List[float]] = [[] for _ in range(self._num_images)]
+
+        for j in range(num_tracks):
+            track = self._sfm_result.get_track(j)
+            num_measurements = track.number_measurements()
+            measurements = [track.measurement(k) for k in range(num_measurements)]
+            w_x = track.point3()
+            for k1 in range(num_measurements):
+                for k2 in range(k1 + 1, num_measurements):
+                    i_a = measurements[k1][0]
+                    i_b = measurements[k2][0]
+
+                    key_a = -1
+                    key_b = -1
+
```

Comment:
it looks like the 6 lines of code below are duplicated, so would a for loop over zip() allow us to reduce copying any code?

Reply:
for example, we could do the following (returning early, instead of nesting : - ) ) should be k1, not j1, also, I believe

```python
for i in i_a, i_b:
    # Check if measurement k belongs to a valid camera
    if not i_a in self._camera_idx_to_patchmatchnet_idx:
         continue
    key = self._camera_idx_to_patchmatchnet_idx[i]
     # Calculate track_i's depth in the camera pose
     z = self._sfm_result.get_camera(i).pose().transformTo(w_x)[-1]
     # Update image i_a's depth list only when i_a in a valid camera
      depths[key].append(z)
```

Reply:
marking as resolved now, thanks for updating this